src/WebJobs.Script.WebHost/Diagnostics/DeferredLogDispatcher.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics
+{
+    public class DeferredLogDispatcher : IDeferredLogDispatcher, IDisposable
+    {
+        private readonly Channel<DeferredLogEntry> _channel = Channel.CreateBounded<DeferredLogEntry>(new BoundedChannelOptions(150)
+        {
+            FullMode = BoundedChannelFullMode.DropOldest,
+            // Avoids locks and interlocked operations when reading from the channel.
+            SingleReader = true,
+            SingleWriter = false
+        });
+
+        private readonly List<ILoggerProvider> _forwardingProviders = new(1);
+        private bool _isEnabled = true;
+        private bool _disposed = false;
+
+        public int Count => _channel.Reader.Count;
+
+        bool IDeferredLogDispatcher.IsEnabled => _isEnabled;
+
+        public void AddLoggerProvider(ILoggerProvider provider)
+        {
+            _forwardingProviders.Add(provider);
+        }
+
+        public void Log(DeferredLogEntry log)
+        {
+            _channel.Writer.TryWrite(log);
+        }
+
+        public void ProcessBufferedLogs(bool runImmediately = false)
```

Comment:
I recommend accepting `IEnumerable<ILoggerProvider>` here for forwarding. Can then remove the instance level `_forwardingProviders` and `AddLoggerProvider`.

Reply:
Removed _forwardingProviders